### PR TITLE
planner: fix index merge plan when expr cannot be pushed to tikv

### LINKED ISSUE #30200: Got error when substring is used in filter
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

    drop table t1;
    create table t1(c1 varchar(100), c2 varchar(100), key(c1), key(c2), c3 varchar(100));
    insert into t1 values('ab', '10', '10');
    explain select /*+ use_index_merge(t1) */ * from t1 where c1 = 'ab' or c2 = '10' and substring(c3, 10) = 10;
    select /*+ use_index_merge(t1) */ * from t1 where c1 = 'ab' or c2 = '10' and substring(c3, 10) = 10;

### 2. What did you expect to see? (Required)
no error

### 3. What did you see instead (Required)
other error: [components/tidb_query_expr/src/lib.rs:705]: ScalarFunction Substring2ArgsUtf8 is not supported in batch mode

### 4. What is your TiDB version? (Required)

master

### PR DESCRIPTION
**Pull request by guo-shaoge**
Signed-off-by: guo-shaoge <shaoge1994@163.com>

### What problem does this PR solve?

Issue Number: close #30200

Problem Summary:  The original implementation of IndexMerge didn't consider situations when there are exprs which cannot be pushed to TiKV.  It will cause two problems:
1. We may get error just like #30200.
2. We have to use TableScan. But actually we can put the `Selection` above IndexMerge. The latter plan is faster in some situations.

### What is changed and how it works?
1. In `generateIndexMergeOrPaths`, we consider `allConds` in DataSource instead of only `pushedDownConds`. So we can generate IndexMerge path even exprs in filter cannot be pushed down to TiKV.
2. In `buildIndexMergeTableScan`, we check `tableFileders` in DataSource explicitly. There will be two situations:
    **2.1**: Filter can be pushed to TiKV: we put these filters in copTask's tablePlan, which will be executed by coprocessor.
    **2.2**: Filter can not be pushed to TiKV, but can be pushed to other storages: we will create a new Selection to handle these filters(_copTask.handleRootTaskConds_). No need to consider expressions that can neither be pushed down tikv nor other engines, because in logical optimization we already put these filters in Selection.(check [here](https://github.com/pingcap/tidb/blob/732b7eb128ed0f251d9ce92509a3aaad63f02a80/planner/core/rule_predicate_push_down.go#L122).

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [x] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
planner: fix index merge plan when expr cannot be pushed to tikv
```

### CODE REVIEW

File: planner/core/integration_test.go [hunk 1]
```diff
@@ -4954,6 +4954,32 @@ func (s *testIntegrationSuite) TestIssue30094(c *C) {
...
 
+func (s *testIntegrationSuite) TestIssue30200(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t1;")
+	tk.MustExec("create table t1(c1 varchar(100), c2 varchar(100), key(c1), key(c2), c3 varchar(100));")
+	tk.MustExec("insert into t1 values('ab', '10', '10');")
+	// lpad has not been pushed to TiKV or TiFlash.
+	tk.MustQuery("explain select /*+ use_index_merge(t1) */ * from t1 where c1 = 'ab' or c2 = '10' and char_length(lpad(c1, 10, 'a')) = 10;").Check(testkit.Rows(
```
Comment (qw4990) [inline planner/core/integration_test.go:4965]: Better to use `explain format=brief ...`.

**Review by qw4990 — APPROVED** [review]

File: planner/core/integration_test.go (hunk 1 above)
Comment (guo-shaoge) [reply planner/core/integration_test.go]: done

File: planner/core/find_best_task.go [hunk 2]
```diff
@@ -1124,17 +1128,40 @@ func (ds *DataSource) buildIndexMergeTableScan(prop *property.PhysicalProperty,
 		ts.stats.StatsVersion = statistics.PseudoVersion
 	}
 	if len(tableFilters) > 0 {
+		// Here we want to add a Selection for exprs that cannot be pushed to TiKV, but some of them has already been put
+		// in Selection above DataSource, so we need to filter these exprs.
+		_, filtersInSelection := expression.PushDownExprs(sessVars.StmtCtx, tableFilters, ds.ctx.GetClient(), kv.UnSpecified)
```
Comment (XuHuaiyu) [inline planner/core/find_best_task.go:1133]: extract a function to extract the functions which can not be pushed down to tikv but  can be pushed down to other storages.

Comment (guo-shaoge) [reply planner/core/find_best_task.go]: done

File: planner/core/find_best_task.go [hunk 3]
```diff
@@ -1124,17 +1127,40 @@ func (ds *DataSource) buildIndexMergeTableScan(prop *property.PhysicalProperty,
...
+			return sel, partialCost, remainedFilters, nil
+		}
+		return ts, partialCost, remainedFilters, nil
+	}
+	return ts, partialCost, nil, nil
+}
+
+// extractFiltersForIndexMerge returns:
+// 1. exprs that can be pushed to TiKV.
+// 2. exprs that cannot be pushed to TiKV but can be pushed to other storages.
```
Comment (XuHuaiyu) [inline planner/core/find_best_task.go:1149]:
```suggestion
// `pushed`: exprs that can be pushed to TiKV.
// `remained`: exprs that can NOT be pushed to TiKV but can be pushed to other storage engines.
```

File: planner/core/find_best_task.go [hunk 4]
```diff
@@ -1124,17 +1127,40 @@ func (ds *DataSource) buildIndexMergeTableScan(prop *property.PhysicalProperty,
...
+// Why: IndexMerge only works on TiKV, so we need to find all exprs that cannot be pushed to TiKV, and add a new Selection above IndexMergeReader.
```
Comment (XuHuaiyu) [inline planner/core/find_best_task.go:1150]:
```suggestion
// Why do we need this func?
// IndexMerge only works on TiKV, so we need to find all exprs that cannot be pushed to TiKV, and add a new Selection above IndexMergeReader.
```

File: planner/core/task.go [hunk 5]
```diff
@@ -1077,8 +1083,6 @@ func (t *copTask) convertToRootTaskImpl(ctx sessionctx.Context) *rootTask {
 		newTask.p = sel
 		sel.cost = newTask.cost()
```
Comment (guo-shaoge) [inline planner/core/task.go:1084]: Here we add a new Selection operator, guess we should update newTask's cost?
`newTask.addCost(t.count() * sessVars.CPUFactor)` just like `PhysicalSelection.attach2Task` did.
@winoros

File: planner/core/find_best_task.go [hunk 6]
```diff
@@ -1124,17 +1127,40 @@ func (ds *DataSource) buildIndexMergeTableScan(prop *property.PhysicalProperty,
...
+// 	But the new Selection should exclude exprs that cannot be pushed to other storages either.
```
Comment (XuHuaiyu) [inline planner/core/find_best_task.go:1151]:
```suggestion
// 	But the new Selection should exclude the exprs that can NOT be pushed to ALL the storage engines.
```

File: planner/core/integration_test.go [hunk 7]
```diff
@@ -4992,6 +4992,47 @@ func (s *testIntegrationSuite) TestIssue30094(c *C) {
...
+func (s *testIntegrationSuite) TestIssue30200(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t1;")
+	tk.MustExec("create table t1(c1 varchar(100), c2 varchar(100), key(c1), key(c2), c3 varchar(100));")
+	tk.MustExec("insert into t1 values('ab', '10', '10');")
+	// lpad has not been pushed to TiKV or TiFlash.
+	tk.MustQuery("explain format=brief select /*+ use_index_merge(t1) */ * from t1 where c1 = 'ab' or c2 = '10' and char_length(lpad(c1, 10, 'a')) = 10;").Check(testkit.Rows(
+		"Selection 8000.00 root  or(eq(test.t1.c1, \"ab\"), and(eq(test.t1.c2, \"10\"), eq(char_length(lpad(test.t1.c1, 10, \"a\")), 10)))",
```
Comment (guo-shaoge) [inline planner/core/integration_test.go:5004]: The estRow of `Selection` looks strange. We should open a new issue.

File: planner/core/find_best_task.go [hunk 8]
```diff
@@ -973,14 +974,17 @@ func (ds *DataSource) convertToIndexMergeScan(prop *property.PhysicalProperty, c
 	if prop.ExpectedCnt < ds.stats.RowCount {
 		totalRowCount *= prop.ExpectedCnt / ds.stats.RowCount
 	}
-	ts, partialCost, err := ds.buildIndexMergeTableScan(prop, path.TableFilters, totalRowCount)
+	ts, partialCost, remainedFilters, err := ds.buildIndexMergeTableScan(prop, path.TableFilters, totalRowCount)
```
[the remark this answers is not in the document]
Comment (guo-shaoge) [reply planner/core/find_best_task.go]: done

File: cmd/explaintest/r/index_merge.result [hunk 9]
```diff
@@ -237,11 +237,11 @@ insert into t1(c1, c2) values(1, 1), (2, 2), (3, 3), (4, 4), (5, 5);
 explain select /*+ use_index_merge(t1) */ * from t1 where c1 < 10 or c2 < 10 and c3 < 10 order by 1;
 id	estRows	task	access object	operator info
 Sort_5	4060.74	root		test.t1.c1
-└─IndexMerge_12	2250.55	root		
-  ├─IndexRangeScan_8(Build)	3323.33	cop[tikv]	table:t1, index:c1(c1)	range:[-inf,10), keep order:false, stats:pseudo
-  ├─IndexRangeScan_9(Build)	3323.33	cop[tikv]	table:t1, index:c2(c2)	range:[-inf,10), keep order:false, stats:pseudo
-  └─Selection_11(Probe)	2250.55	cop[tikv]		or(lt(test.t1.c1, 10), and(lt(test.t1.c2, 10), lt(test.t1.c3, 10)))
-    └─TableRowIDScan_10	5542.21	cop[tikv]	table:t1	keep order:false, stats:pseudo
+└─Selection_12	2250.55	root		or(lt(test.t1.c1, 10), and(lt(test.t1.c2, 10), lt(test.t1.c3, 10)))
```
Comment (guo-shaoge) [inline cmd/explaintest/r/index_merge.result:240]: Why `Selection_12` not push down: `c3` is generated column.

**Review by XuHuaiyu — APPROVED** [review]
LGTM

**Pull request merged**